### ISSUE TITLE
Volumes need to be defined in docker-compose

### ISSUE THREAD

**Issue by bnhf** [gregewing/apcupsd#5]
@gregewing 

Rather than tacking-on to the PR that didn't make the grade, I thought I'd open an issue to expand on a couple of topics:

Following the direction you want to go, and for the purposes of the README with the recently updated docker-compose, the volume needs to be defined in the yaml. It does not need to be created in advance. Also, the name of the stack is appended to the front of the volume, so a volume name of "config" results in a directory named "apcupsd_config" in /var/lib/docker/volumes (assuming a stack name of apcupsd) :

```yml
version: '3.7'
services:
  apcupsd:
    image: gregewing/apcupsd:latest
    container_name: apcupsd
    devices:
      - /dev/usb/hiddev0 # This device needs to match what the APC UPS on your system uses.
    ports:
      - 3551:3551
    environment:
      - TZ=${TZ}
    volumes:
      - /var/run/dbus/system_bus_socket:/var/run/dbus/system_bus_socket
      - config:/etc/apcupsd
    restart: unless-stopped
volumes:
  config:
```
And this results in the following directory being created, and populated:

![screenshot-raspberrypi10-2023 02 02-19_05_34](https://user-images.githubusercontent.com/41088895/216495252-752d7f8f-3456-49c4-b3f8-66815abd05c3.png)

I think you'll find that the start.sh in the PR works equally well whether the binding is to a directory (like /data/apcupsd) or to a volume (like config:). Either way, somebody can launch an apcupsd daemon container all from a Portainer-Stack, without having to execute or modify anything from the command line.

If you're willing, I'd like to figure out why you were unsuccessful using the docker-compose from my fork. I've been using it for my "Ultimate APC UPS Monitoring" project, and it's been dead-simple and reliable so far. Did your portainer stack look like this?:

![screencapture-raspberrypi10-9000-2023-01-30-14_29_02](https://user-images.githubusercontent.com/41088895/216494825-cdd853cd-8140-4462-b41a-49ff4840fdf8.png)

**Comment by gregewing** [comment]
Well, admittedly I don't use Docker Compose, and I tested your fork by creating a container manually and exposing to it the various resources and a blank volume.  Yes, the volume gets populated, but if I remove and recreate the container any customizations that I have made in them are over-written, which is far from ideal for wen a new release of the container becomes available and watchtower does its thing.  Is suspect this is a result of your start.sh script.

We should not be prescriptive about how a container is created, so it should work the same way regardless of how the container is created or managed (docker run / portainer / docker compose / etc.)

I'll work on the job of using Docker Environment variables to push configs to the apcupsd.config file, but I'm busy with other projects and starting a new job too.  I know how to do it in a way that will be non-destructive and have it persist a recreation of the container while re-using an existing mounted folder or volume.

**Comment by bnhf** [comment]
@gregewing 

I spent a few minutes this morning looking at this, and it turns out that I hadn't rebuilt the image on Docker Hub with the changes I submitted in the PR. The start.sh in the Docker Hub container was an earlier version, where I was moving the files unconditionally. Anyway, the version in the PR was correct (with the addition of conditional moves for hosts.conf and doshutdown as referenced  in the PR comment).

I've rebuilt the multi-arch container on Docker Hub now, as I know you'll want to do some additional testing. I've tested it using   a docker volume, where on the first run I set the UPSNAME environment variable -- then stopped the stack and removed the environment variable. apcupsd.conf was not overwritten on the next run, evidenced by the UPSNAME setting in the file remaining in its previously modified state.

Apologies for not updating my Docker Hub container with the latest script -- the fact that apcupsd.conf was getting overwritten on each container run, wasn't obvious on my end since my desired changes (via the environment variables) were also getting performed on each run.

**Comment by bnhf** [comment]
@gregewing 

Me again. :-) OK, so I've officially tested every bind mount scenario I could think of, and they're all working with bnhf/apcupsd:latest. There are differences in the docker-compose (just as there are from the command line) depending on which approach you use, directory binding, new volume binding or existing volume binding. These differences only come into play when there are no existing config files -- after the files are created, they will persist and be used, unless specifically deleted.

Here's the "universal" docker-compose with comments regarding first-run options:

```yml
version: '3.7'
services:
  apcupsd:
    image: bnhf/apcupsd:latest
    container_name: apcupsd
    devices:
      - /dev/usb/hiddev0 # Check the /dev/usb folder on the UPS connected system to determine your options here
    ports:
      - 3551:3551
    environment:
      - UPSNAME=${UPSNAME} # This name will be used in a number of monitoring applications. 1 to 8 characters, no spaces
      - UPSCABLE=${UPSCABLE} # Default is usb
      - UPSTYPE=${UPSTYPE} # Default is usb
      - TZ=${TZ} # Local timezone of the UPS device
    volumes:
      - /var/run/dbus/system_bus_socket:/var/run/dbus/system_bus_socket
      - /data/apcupsd:/etc/apcupsd # /etc/apcupsd can be bound to a directory or a docker volume
    restart: unless-stopped
# volumes: # Use this section for volume bindings only.
#   config: # The name of the stack will be appended to the beginning of this volume name, if the volume doesn't already exist
#           # If you created the docker volume in advance, use that name here, and for the /etc/apcupsd binding above
#     external: true # Use this directive if you created the docker volume in advance
```

Also, I added some basic log output to indicate whether existing files are being used, and if apcupsd.conf variables are being mofified:

![screenshot-raspberrypi10-2023 02 03-20_38_29](https://user-images.githubusercontent.com/41088895/216746127-08c1c815-9711-491e-b398-476fb2a948be.png)

**Comment by gregewing** [comment]
As I do't use Docker Compose, i'm gong to leave it to you to make sure I have teh right syntax and content in the ReadMe.  but please can we come back to that on a little bit?  I want to settle ENVs first.  

First I wanted to let you know that I'm working on integrating your changes for ENV variables, initially bt defining the ENVs in the Dockerfile,  that way they are defined in the container when it is downloaded.  I've done this now and pushed all teh variants based on the same Dockerfile.  Please can you go in and check the current set of images, and test teh following use-cases:

- Does the presence of the ENVs affect anything with a running container that does not use the ENVs to manage the configuration of apcupsd?
- Does the presence of ENVs in the container impact any defined ENVs in yor docker compose YAMLs?
- If you redeploy a container using Docker Compose, do the ( default empty ) ENVs in the container affect anything with the docker compose ?

I've left the ENVs defaulting to empty so that the next step does not immediately impact anyone using this container with old-fashioned config files, and are not keeping an eye on the documentation.   The idea being that any changes you or I make are non-destructive to existing containers, especially when they upgrade by way of Watchtower or whatever.

While I'm waiting for you to feedback to me on the ENVs, I'll get on with integrating your ideas for a start script to replace just running the daemon.  I'll do this just a little bit differently than your example, but you will recognize a lot of your code in there ;-)

**Comment by bnhf** [comment]
@gregewing 

Thanks for taking time to work on this!

I'm trying to understand the advantage of defining these empty environment variables in the Dockerfile. Since we're using them to modify configuration files via a startup script, and there are no default values that aren't already specified in the several .conf files. I can see in other scenarios where this would be useful -- but maybe this will become clear in what you have in mind for start.sh.

I've done a basic test so far using what you just pushed to Docker Hub, and I don't think any issues are likely. In fact the empty variables appear to have no impact on the environment:

```console
root@236265dfeecb:/# printenv
HOSTNAME=236265dfeecb
PWD=/
TZ=US/Mountain
HOME=/root
LANG=C.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
TERM=xterm
SHLVL=1
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
DEBIAN_FRONTEND=noninteractive
_=/usr/bin/printenv
root@236265dfeecb:/#
```

**Comment by gregewing** [comment]
The only advantage is that when starting the container in a non-compose environment, we don't have to then manually add them to Portainer, the show up ready to be filled by the used.
________________________________

**Comment by bnhf** [comment]
@gregewing 

I haven't pushed my latest stuff to GitHub yet, but I wanted to give you a preview of a few concepts I've been moving towards:

- In order to support both directory bind mounts, and volume bind mounts, we need to move the contents of /etc/apcupsd to a holding directory. I believe this directory should be changed from /usr/local/bin (which is on the PATH), to /opt/apcupsd (which is a "standard" directory for third party apps, and is not on the PATH). Further, I think we should use cp instead of mv, to get files into the bound /etc/apcupsd directory if they don't already exist. Since the original files won't be on the PATH, they can safely be left in /opt/apcupsd. This works better if the container is restarted, as the original files still exist when start.sh is run again.
- The only file in the /scripts directory that actually needs to be there is doshutdown. The other two: apcupsd.conf and hosts.conf don't contain any changes of consequence. The two COPY commands in the Dockerfile should be moved after RUN, so that the contents of /scripts replaces files of the same name created when apcupsd is installed.
- The number of if-then file checks I'm doing has grown, so it'll be tidier to put them in an array. This can be done for the environment variables as well. It doesn't really make much difference other than being easier to read and modify. So for example, here's the array version of the file existence checks in /etc/apcupsd:
```bash
# Check if /etc/apcupsd files exist, and copy them from /opt/apcupsd if they don't
files=( apcupsd.conf hosts.conf doshutdown apccontrol changeme commfailure commok killpower multimon.conf offbattery onbattery ups-monitor )

for i in "${files[@]}"
  do
    if [ ! -f /etc/apcupsd/$i ]; then
      cp /opt/apcupsd/$i /etc/apcupsd/$i \
      && echo "No existing $i found"
    else
      echo "Existing $i found, and will be used"
    fi
  done
```
I'd like to be able submit some of this using the PR process, but our code bases have gone in somewhat different directions. Hopefully we can get them synchronized again soon, as I hate for us to be duplicating effort.

**Comment by gregewing** [comment]
i'm trying to get caught up with your changes, but you are moving quicker than I am able to keep up.  Perhaps you can give me a little time to wrap my head around all your proposed changes and get them into the container in a format that I'm happy with.  Then we can sort out what comes next.

In short I think a bit of organization is necessary so that we can make the best use of our time.  Therefore I propose the following list of objectives, each one to be completed to our mutual satisfaction before moving on to the next, the key being to work on a single file or feature at a time:


  1.  Greg to update the container with a entrypoint.sh script based on the work Scott has done to date, specifically to integrate the use of environment variables defined at runtime outside of the container ( any number of mechanisms) and make it such that these environment variables only affect a deployed container if they are non-blank, i.e. set specifically by the administrator.  This should over-ride whatever content is in the apcupsd.conf file, but non-destructively so that if the user deletes or empties the ENV Variable (say, in Portainer) then the default behaviour is restored.  I'll base this on the array that you provided.
  2.  Scott to update the README to represent the changes required for docker-compose use of the container with the features above.
  3.  Scott to propose, by way of Pull Request, the specific changes required to implement hosts.conf and master/slave with configuration done entirely outside the container using ENV Variables.
  4.  Scott to update the README to represent the changes required for docker-compose use of the container with the features above
  5.  Look at integrating apcupsd-cgi into the container.
  6.  any
  7.  other
  8.  ideas
  9.  ?

Maybe we log these as specific feature requests so we can track them?


________________________________

**Comment by bnhf** [comment]
@gregewing 

I'm with you -- we need to get organized. At this point if we can get the start.sh script integrated, and update the README to reflect the new variables, then I can sync my code base to yours. After that, I'll submit small PRs, which you can accept, reject or modify. And yes, let's wait on spinning up the whole master/slave capability until we can do it by PR.

Here's what I hope will be the final start.sh (for the moment) with the environment variables set through an array. This should work well as it's compact, and easy to modify as far as adding files to /etc/apcupsd, or adding environment variables. There's a one-liner for copying the /etc/default/apcupsd file, and a one-liner to fix a second instance of UPSNAME in apcupsd.conf. So basically one spot for any future files tweaks, and one spot for environment variable tweaks -- things that don't lend themselves to being handled in an array:

```bash
#! /bin/bash

# Config file moves transferred from Dockerfile to support
# binding /etc/apcupsd to user-specified host directory

# Copying the file that contains "ISCONFIGURED=yes" to the /etc/default directory
cp /opt/apcupsd/apcupsd /etc/default/apcupsd

# Check if /etc/apcupsd files exist, and copy them from /opt/apcupsd if they don't
files=( apcupsd.conf hosts.conf doshutdown apccontrol changeme commfailure commok killpower multimon.conf offbattery onbattery ups-monitor )

for i in "${files[@]}"
  do
    if [ ! -f /etc/apcupsd/$i ]; then
      cp /opt/apcupsd/$i /etc/apcupsd/$i \
      && echo "No existing $i found"
    else
      echo "Existing $i found, and will be used"
    fi
  done

# A one-liner to add an extra # to the second UPSNAME used for EPROM updates, to keep it from being further changed by sed
sed -i 's/^#UPSNAME UPS_IDEN/##UPSNAME UPS_IDEN/' /etc/apcupsd/apcupsd.conf

# Check if environment variables are set, and if so update apcupsd.conf
settings=( "UPSNAME" "UPSCABLE" "UPSTYPE" "DEVICE" "POLLTIME" "NETSERVER" "NISIP" "NISPORT" "ONBATTERYDELAY" "BATTERYLEVEL" "MINUTES" "TIMEOUT" "SELFTEST" )

for i in ${settings[@]}
  do
    if [ ! -z ${!i} ]; then
      sed -i -r 's/(^'"$i"'.*|^#'"$i"'.*)/'"$i"' '"${!i}"'/' /etc/apcupsd/apcupsd.conf \
      && awk '$1 ~ /^'"$i"'/' /etc/apcupsd/apcupsd.conf
    fi
  done

/sbin/apcupsd -b
```

Not a lot of code, but it takes care of numerous tasks. :-)

**Comment by gregewing** [comment]
I struggled with getting my head round why we need to do a file copy in the startup script again. Can you please explain it slowly for me again, like I’m 5. I keep coming back to the argument that docker will put files in the bind mount or folder or even in the right place of the user doesn’t use a bind of any type, at the point when the container is created, which is always the case. Stopping and starting a container won’t do this, but a rip and recreate will every time. I just can’t see a need for it, so please can you explain it to me again ?

Thanks and Regards
Greg Ewing
________________________________

**Comment by bnhf** [comment]
@gregewing 

> I struggled with getting my head round why we need to do a file copy in the startup script again.

The best example I can think of is that we're doing it for the new user that wants to bind mount a directory to /etc/apcupsd. And, let's say that directory doesn't already exist. Docker creates the directory, but it's always created empty. That empty directory is then bound to /etc/apcupsd in the container, the result being that there are no data files in /etc/apcupsd.

If you were to launch the container again, without any bind mount, the files would magically reappear in /etc/apcupsd in the container. They were there all along, but hidden behind the bind mount to a directory on the host.

Obviously this isn't the way docker volumes work, but realistically, most people don't use volumes. As I recall, volumes are a newer feature of Docker, so people got in the habit of using directory binding and putting their container data in a place that's convenient for them -- I typically use /data.

So to summarize, by doing the copies after /etc/apcupsd is bind mounted to whatever, we can be confident that the files we want in that directory will be there for a new user. And by doing it conditionally, we're also good with the user that mounts a directory or volume where the data files are already present.

**Comment by gregewing** [comment]
Ok, thanks for that.

I did a little experimentation with Docker Compose (Portainer Stacks) and I see the behaviour that you are describing.

However, I deeply disagree that "most people don't use volumes".  Volumes are the preferred mechanism for persisting data generated by and used by Docker containers. While bind mounts are dependent on the directory structure of the host machine, volumes are completely managed by Docker. Volumes are often a better choice than persisting data in a container’s writable layer, because a volume does not increase the size of the containers using it, and the volume’s contents exist outside the lifecycle of a given container.

Yes, it comes down to personal preference, and I generally don't advocate forcing people to change to accommodate a technology, ideally it should be the other way round. Technology should be able to meet the varying demands of the user, without the user having to spend too much time thinking about it.

That said, it is my opinion that this a niche use-case, and an inefficient one at that.  I don't think we should be altering the container to try and support it.   It is more efficient to move the files into place within the Dockerfile, so they are there at deploy time:  We do it once, so we don't have to do it a hundred thousand times.  I think we should document this in the Readme and encourage people to use volumes as follows if they want to use docker compose:

version: '3.7'
services:
  apcupsd:
    image: gregewing/apcupsd:latest
    container_name: apcupsd-stack
    devices:
      - /dev/usb/hiddev0
    ports:
      - 3551:3551
    environment: # Delete or comment out any environment variables you don't wish to change
      - UPSNAME=${UPSNAME} # This value will display in apcupsd-cgi details.
      - UPSCABLE=${UPSCABLE} # Default value is usb
      - UPSTYPE=${UPSTYPE} # Default value is usb
      - DEVICE=${DEVICE} # Default value is <blank>
      - NETSERVER=${NETSERVER} # Default value is on
      - NISIP=${NISIP} # Default value is 0.0.0.0
      - TZ=${TZ} # Default value is Europe/London
    volumes:
      - /var/run/dbus/system_bus_socket:/var/run/dbus/system_bus_socket
      - /etc/apcupsd
    restart: unless-stopped


This creates a container and dynamically creates a volume and puts the configuration files in the right place inside the container.  They are also accessible on the host in the volume folder on the filesystem on the host in /var/lib/docker/volumes/, so if someone really wants or needs to alter a config file without entering the container, or even with the container stopped, then it can be done by navigating to the correct folder and doing it there.

So, I'm sorry to say, unless you can convince me that there is a cast-iron need to put the config file management into the startup script, then I'm going to reject the idea, and propose that those users who feel they might benefit from that approach, take a different approach to managing their persistent data.  I think they will thank me in the long-run.

Of course, you are more than welcome to fork the container and maintain a copy yourself based on the one I publish, or create a new one all together and put that functionality in if it's something that you feel strongly about. You can do this easily by setting my container as the FROM statement in your Dockerfile, or by copying and altering the content of my Dockerfile in your own project for your own purposes.


I'm very happy to continue working with you on adding the ENV variable integration as we have been discussing as I see this as a great value-add to the container, and will happily acknowledge the valuable input you have provided.


Greg
________________________________

**Comment by bnhf** [comment]
@gregewing 

> Volumes are the preferred mechanism for persisting data generated by and used by Docker containers

I don't necessarily disagree with that. Volumes have plenty of advantages, however as you said in one of your previous messages (paraphrasing here): we shouldn't be prescriptive about how people deploy.

> That said, it is my opinion that this a niche use-case, and an inefficient one at that. I don't think we should be altering the container to try and support it.

Definitely not niche. I'm about as close to 100% Docker as you can get on my Linux systems, and I generally use suggested docker-compose ymls when available. Looking at the dozen plus systems I have running Docker, there were only a few applications that were using volumes. Portainer itself is the big one, and that's present on all of my systems -- but otherwise the only other one I could find is Paperless-ngx.

> Of course, you are more than welcome to fork the container and maintain a copy yourself based on the one I publish, or create a new one all together and put that functionality in if it's something that you feel strongly about.

Well shoot, I was hoping we'd see a little more eye-to-eye on this. My hope was that we could create something that would work for both us, but it's your project and I get that. Open source can be especially amazing through collaboration. Not having to do everything yourself, and being able to learn through the input of others is something I'm a great believer in.

Let me know if you change your mind, but otherwise I'll probably move in the direction of spinning my fork off into its own repository.

Sorry this didn't work out, but the best of luck to you in the future, and with your new job.

**Comment by gregewing** [comment]
Thanks for your understanding. I sense the disappointment.

You know you can share volumes between running containers, right?  Just need to be careful with read/write privileges, as you would with file system bind mounts.

Thanks and Regards
Greg Ewing
________________________________
